MVP here we come

CBroz1 opened the pull request:
### Commits
- Minor PEP8 changes at [commit aa4961e](https://github.com/CBroz1/element-deeplabcut/commit/aa4961ea2a7fcc6a375edddaee22562f981a8fef)
- Major changes at [commit cef65ed](https://github.com/CBroz1/element-deeplabcut/commit/cef65ed8d4dc27ee672f887baa704c49cb5e048f), listed below
- See workflow [demo notebook](https://github.com/CBroz1/workflow-deeplabcut/blob/dev/notebooks/1_Explore_Workflow.ipynb)

### Changes
- README: add note on dependency conflict numpy==1.2
- dlc.py:
   - remove arg for get_dlc_processed_dir
   - Add note on scorer_legacy
   - add docstrings
   - add note on continue from prev snapshot
   - add set shuffle/trainindex/maxiters to int for model training make
   - BodyPart.insert_new - refactor logic for yaml type check
   - Model.insert_new_model - refactor config yaml type check
       - Add scorer legacy check
       - add dlc version insertion
       - add body part description arg
   - ModelEvaluation - make items nullable bc saw eval that didn't return some value
       - refactor config yaml type check
   - PoseEstimationTask - permit estimation paramset to pass to analyze_videos
       - Type set to string for concatenating output dir
       - add insert helper function
   - PoseEstimation - permit output_dir as fullpath, per 'relative' arg in task table
       - permit passing analyze_videos params
- dlc_reader:
   - add conditional logic for if passed the output dir instead of project dir
   - assumes output dir is under project dir
   - do_pose_estimation - permit args passed from PoseEstimationTask.params

CBroz1 committed:
Merge from CBroz1/dev

Recent work

CBroz1 committed:
Update element_deeplabcut/dlc.py

Co-authored-by: Thinh Nguyen <thinh@vathes.com>

CBroz1 committed:
Merge pull request #3 from CBroz1/dev

Add functions to link tables

CBroz1 committed:
Merge pull request #4 from ttngu207/main

prototype revision no 2

CBroz1 committed:
PEP8

CBroz1 committed:
See details - long list

README: add note on dependency conflict numpy==1.2

dlc schema:
- remove arg for get_dlc_processed_dir
- Add note on scorer_legacy
- add docstrings
- add note on continue from prev snapshot
- add set shuffle/trainindex/maxiters to int for model training make
- BodyPart.insert_new - refactor logic for yaml type check
- Model.insert_new_model - refactor config yaml type check
    - Add scorer legacy check
    - add dlc version insertion
    - add body part description arg
- ModelEvaluation - make items nullable bc saw eval that didn't return some value
    - refactor config yaml type check
- PoseEstimationTask - permit estimation paramset to pass to analyze_videos
    - Type set to string for concatenating output dir
    - add insert helper function
- PoseEstimation - permit output_dir as fullpath, per 'relative' arg in task table
    - permit passing analyze_videos params

dlc_reader:
- add conditional logic for if passed the output dir instead of project dir
- assumes output dir is under project dir
- do_pose_estimation - permit args passed from PoseEstimationTask.params

CBroz1 committed:
update diagram

ttngu207 commented on element_deeplabcut/dlc.py:
```diff
@@ -129,27 +129,42 @@ class File(dj.Part):
     @classmethod
-    def insert_new_params(cls, paramset_desc: str, params: dict, paramset_idx: int = None):
+    def insert_new_params(cls, paramset_desc: str, params: dict,
+                          paramset_idx: int = None, skip_duplicates=False):
```
why do we need `skip_duplicates` here? 
Duplicated entries are already handled in this function
+ if the new entry is a duplicate that is identical to existing entry, this would skip
+ if the new entry has a duplicated name but different parameters, errors will be throw
+ if the new entry has different name but identical parameters, a informative statement will be printed

CBroz1 replied:
I'll revert this. Not sure why I added it. I was seeing the assertion warning when testing the ingest function from the workflow, but we should flag this case of 'same parameters, new name'

ttngu207 commented on element_deeplabcut/dlc.py:
```diff
@@ -209,6 +228,8 @@ def make(self, key):
         dlc_config = (ModelTrainingParamSet & key).fetch1('params')
         dlc_config['project_path'] = project_path.as_posix()
         dlc_config['modelprefix'] = model_prefix
+        dlc_config['train_float'] = dlc_config['TrainingFraction'
```
Should we rename `train_float` to `train_fraction`, for consistency?

CBroz1 replied:
Sure

ttngu207 commented on element_deeplabcut/dlc.py:
```diff
@@ -251,28 +275,44 @@ def make(self, key):
+        if new_body_parts is not None:  # Required bc np.array is ambiguous as bool
+            print(f'Existing body parts: {tracked_body_parts}')
+            print(f'New body parts: {new_body_parts}')
+            if descriptions:
```
What would happen if `descriptions` is not provided - i.e. equals to `None`?

ttngu207 commented on element_deeplabcut/dlc.py:
```diff
@@ -251,28 +275,44 @@ def make(self, key):
+                print(f'New descriptions: {descriptions}')
+            if dj.utils.user_choice(f'Insert {len(new_body_parts)} new body '
+                                    + 'part(s)?') == 'yes':
+                cls.insert([{'body_part': b, 'body_part_description': d}
+                           for b, d in zip(new_body_parts, descriptions)])
```
If users do not provide `descriptions`, this line would fail?

CBroz1 replied:
It needs another conditional before L312 to work. Whether or not to add depends on discussion above.
```python
if descriptions is None:
    descriptions = ["" for x in range(len(new_body_parts))]
```

ttngu207 commented on element_deeplabcut/dlc.py:
```diff
@@ -251,28 +275,44 @@ def make(self, key):
+            print(f'Existing body parts: {tracked_body_parts}')
+            print(f'New body parts: {new_body_parts}')
+            if descriptions:
+                assert len(descriptions)\
+                        == len(new_body_parts), ('Descriptions list does not match '
+                                                 + ' the number of new_body_parts')
```
I think this assertion will fail most of the time. Because the `len(new_body_parts)` will not match the `len(descriptions)`
In our function, we have logic for checking which body-part from the config file are new, and only insert those new ones - the `new_body_parts`. In the logic here, we are expecting the users to know, at the time they call this function, what are the new body-parts, and in which order, so they can provide a meaningful `descriptions`.
This is a bit unrealistic and error prone, I'd suggest not even requiring `descriptions` here. The body_part description can be updated after the fact fairly easily and less error prone.

CBroz1 replied:
My understanding was that the elements should model best practices, which, for database integrity, should not include updates; is that the case?

I wanted to provide a mechanism to use this function with descriptions. My expected use-case, while not ideal, was that a user would run once, see the list, generate descriptions, and then run again. Should this helper function instead be separated into two, one to return the list of new items, and a second to insert (optionally with descriptions)?

ttngu207 replied:
Generally we do want to avoid updating, but for metadata at this level, updating is quite common, especially updating a description field (it would be another conversation if we're updating fields such as sampling rate for example). I would rather allow for updating than introducing more complex logic and usage requirement.

ttngu207 commented on element_deeplabcut/dlc.py:
```diff
@@ -296,26 +336,40 @@ class Model(dj.Manual):
-                         shuffle, trainingsetindex, model_description='', model_prefix=''):
+    def insert_new_model(cls, model_name: str, dlc_config: dict, *, shuffle: int,
+                         trainingsetindex, model_description='', model_prefix='',
+                         body_part_descriptions: list = None, paramset_idx: int = None):
```
same as my comment above, let's remove `body_part_descriptions` here altogether

ttngu207 commented on element_deeplabcut/dlc.py:
```diff
@@ -296,26 +336,40 @@ class Model(dj.Manual):
+        :param paramset_idx: index from the ModelTrainingParamSet table
+        """
         from deeplabcut.utils.auxiliaryfunctions import GetScorerName
+        from deeplabcut import __version__ as dlc_version
```
`dlc_version` here is a bit misleading. With this logic, we are storing the `dlc_version` of the python environment at the time of inserting this new model. There is no guarantee that the model itself is actually trained using this same `dlc_model`.
To get a meaningful `dlc_model`, we can either make it an optional input argument, i.e. users would have to tell us which dlc model was used to train the model that they're about to insert (I doubt if most users would have this info available), or we omit `dlc_version` altogether

ttngu207 replied:
An alternative is to read the `dlc_version` from the config yaml file, if it's available there

CBroz1 replied:
Version is not available in the `config.yaml`, no. We cant be certain that the same DLC version is being across network training and DataJoint ingestion, no. It seems unlikely to me that one would update between those steps. 

I'll remove this as a variable in the Model table. Any objections to using the python environment dlc_version to infer legacy naming? If the user's environment is old enough, or if the user specifies in training parameters, we infer a different naming scheme for `dlc_scorer`

ttngu207 commented on element_deeplabcut/dlc.py:
```diff
@@ -296,26 +336,40 @@ class Model(dj.Manual):
+                                           pathlib.Path(dlc_config))
+            assert dlc_config_fp.exists(), ('dlc_config is neither dict nor filepath'
+                                            + f'\n Check: {dlc_config_fp}')
+            if dlc_config_fp.suffix in ('.yml', '.yaml'):
```
need an accompanying `else` here

CBroz1 replied:
I'll change the L369 assertion to cover both that the path exists and that it specifies a `y*ml`

ttngu207 commented on element_deeplabcut/dlc.py:
```diff
@@ -362,42 +426,49 @@ def insert_new_model(cls, model_name: str, dlc_config: dict, *,
             with cls.connection.transaction:
                 cls.insert1(model_dict)
-                BodyPart.insert_from_config(dlc_config)
+                BodyPart.insert_from_config(dlc_config,
+                                            descriptions=body_part_descriptions)
```
```suggestion
                BodyPart.insert_from_config(dlc_config)

```

ttngu207 commented on element_deeplabcut/dlc.py:
```diff
@@ -362,42 +426,49 @@ def insert_new_model(cls, model_name: str, dlc_config: dict, *,
         import csv
         from deeplabcut import evaluate_network
-        from deeplabcut.utils.auxiliaryfunctions import GetEvaluationFolder
+        from deeplabcut.utils.auxiliaryfunctions import GetEvaluationFolder as GetFolder
```
Why the need to rename `GetEvaluationFolder` to `GetFolder`? Unless there is a naming collision, I think renaming here would make the code less clear, less readable

CBroz1 replied:
I changed because it became very difficult to comply with PEP8 line length limits on L467 with a long function name. I'll revert.

ttngu207 commented on element_deeplabcut/dlc.py:
```diff
@@ -476,23 +568,26 @@ class BodyPartPosition(dj.Part):
+                                         ).fetch1('task_mode', 'pose_estimation_params',
+                                                  'pose_estimation_output_dir')
+        if not Path(output_dir).exists():
+            output_dir = find_full_path(get_dlc_root_data_dir(), output_dir)
```
```suggestion
        output_dir = find_full_path(get_dlc_root_data_dir(), output_dir)
```
There is no need for that `if`

ttngu207 commented on element_deeplabcut/dlc.py:
```diff
@@ -511,12 +606,12 @@ def make(self, key):
     @classmethod
     def get_trajectory(cls, key, body_parts='all'):
```
I'm wondering if we need this method at all. This is a convenient method for data extraction from the pipeline?

CBroz1 replied:
I'm in favor of keeping, to open the door to using pandas for visualization and other preprocessing tools. I expect there will be demand for a tool that lists occurrences or acceleration to get the timing of lever pulls, for example. Seeing the outcome helps show the utility of the schema.

ttngu207 commented on element_deeplabcut/readers/dlc_reader.py:
```diff
@@ -9,16 +9,18 @@
 
     def __init__(self, dlc_dir=None, pkl_path=None, h5_path=None, yml_path=None,
-                 filename_prefix=''):
-        if dlc_dir is None:
+                 filename_prefix='', output_dir=None):
```
Perhaps it was a bit unclear, but `dlc_dir` here is meant to be the same as the `output_dir` you just added - i.e. the directory containing the pose estimated results.
We don't need another `output_dir`, perhaps we can rename `dlc_dir` to `output_dir`, but to me `output_dir` is this context is vague and ambiguous as well. I'm happy to consider a better term here.

CBroz1 replied:
My current intention is for the `dlc_dir` to be the main DLC project directory with relevant subfolders (e.g., `dlc-models`, `videos`). 

Previously, L21 ran into issues because the `dlc.PoseEstimation.populate()` method ran `dlc_reader.PoseEstimation()` with `pose_estimation_output_dir` as the only argument. When this uses our inferred `device_{}_recording_{}_model_{}` output directory, 'videos' is neither the stem nor a subdirectory, it's the parent. 

Do you have different terminology you'd like to use to accommodate this case?

ttngu207 commented on element_deeplabcut/readers/dlc_reader.py:
```diff
@@ -43,11 +45,14 @@ def __init__(self, dlc_dir=None, pkl_path=None, h5_path=None, yml_path=None,
+            else:
+                yml_paths = list(self.dlc_dir.parent.glob(f'{filename_prefix}*.y*ml'))
+                yml_paths = [val for val in yml_paths
+                             if not val.stem == "dlc_config_file"]
```
I'm unclear on the intention of this block, perhaps this is also stemming from the misunderstanding that `dlc_dir` and `output_dir` is meant to be the same thing

ttngu207 commented on element_deeplabcut/readers/dlc_reader.py:
```diff
@@ -157,16 +160,20 @@ def do_pose_estimation(video_filepaths, dlc_model, project_path, output_dir,
     # This is important for parsing the DLC in datajoint imaging
     output_dir.mkdir(exist_ok=True)
-    dlc_cfg_filepath = output_dir / 'dlc_config_file.yaml'
+    dlc_cfg_filepath = dlc_project_path / 'dlc_config_file.yaml'
```
Where is `dlc_project_path` coming from? I'm assuming you mean `project_path` from the input argument? If so, why `project_path`, the intention is to store the pose estimation results at the directory specified by `output_dir`, right?

CBroz1 replied:
`dlc_project_path` is defined in [L150](https://github.com/ttngu207/element-deeplabcut/blob/main/element_deeplabcut/readers/dlc_reader.py#L150) using the `project_path` yes. 

Do we ever read this file? Is it just metadata to store with the output? It's not clear to me why we're keeping [only this restricted set](https://github.com/ttngu207/element-deeplabcut/blob/main/element_deeplabcut/dlc.py#L349-L351)

ttngu207 commented on element_deeplabcut/dlc.py:
```diff
@@ -424,36 +496,56 @@ class PoseEstimationTask(dj.Manual):
-    pose_estimation_output_dir='': varchar(255)  # output directory relative to the root data directory
+    task_mode='load' : enum('load', 'trigger')  # load results or trigger computation
+    pose_estimation_output_dir='': varchar(255) # output dir relative to the root dir
+    pose_estimation_params       : longblob     # analyze_videos params, if not default
```
if we are to allow for users to specify this, we must allow this to be nullable
```suggestion
    pose_estimation_params=null   : longblob     # analyze_videos params, if not default

```

CBroz1 replied:
Sure, adding this to my most recent commit

ttngu207 commented on element_deeplabcut/dlc.py:
```diff
@@ -476,23 +568,26 @@ class BodyPartPosition(dj.Part):
-
-        output_dir = find_full_path(get_dlc_root_data_dir(), output_dir)
-
+        task_mode, params, output_dir = (PoseEstimationTask & key
+                                         ).fetch1('task_mode', 'pose_estimation_params',
+                                                  'pose_estimation_output_dir')
```
Suggesting to rename `params` -> `analyze_video_params`.
And since this is nullable, we need to handle the null case
```
analyze_video_params = analyze_video_params or {}
```

ttngu207 reviewed (changes requested):
I've added a number of comments/suggestions, please review

Commit:
Rebasing

PEP8
README: add note on dependency conflict numpy==1.2
/images/: update diagram
dlc schema:
- remove arg for get_dlc_processed_dir
- Add note on scorer_legacy
- add docstrings
- add note on continue from prev snapshot
- add set shuffle/trainindex/maxiters to int for model training make
- add `TrainingVideo` table for training instead of `VideoRecording`
    - `LabeledFrame` -> `File` generalizable part table
- BodyPart.insert_new
    - refactor logic for yaml type check
    - separate BodyPart functions - return list then insert
- Model.insert_new_model - refactor config yaml type check
    - Add scorer legacy check
    - add dlc version insertion
- ModelEvaluation - make items nullable bc saw eval that didn't return some value
    - refactor config yaml type check
- PoseEstimationTask - permit estimation paramset to pass to analyze_videos
    - Type set to string for concatenating output dir
    - add insert helper function
- PoseEstimation - permit output_dir as fullpath, per 'relative' arg in task table
    - permit passing analyze_videos params
    - permit null analyze_video_param
dlc_reader:
- prev had separate output dir and project dir... removing
- do_pose_estimation - permit args passed from PoseEstimationTask.params
- adjust yml_path in reader to remove dlc_dir.parent
Apply suggestions from code review

CBroz1 committed:
Apply suggestion from code review

Co-authored-by: Thinh Nguyen <thinh@vathes.com>

CBroz1 committed:
additional code review suggestions

CBroz1 reviewed (commented):
Had some follow-up questions

CBroz1 committed:
null analyze_video_param handling

CBroz1 committed:
separate BodyPart functions - return list then insert

CBroz1 commented:
Ran through full pipeline [here](https://github.com/CBroz1/workflow-deeplabcut/blob/dev/notebooks/03-Process.ipynb) - Made some tweaks to resolve `dlc_reader` issue, where DLC expects the project path when generating pose estimation.

CBroz1 committed:
edit README

Commit:
Merge branch 'main' into dev
